src/Interpreters/TraceCollector.cpp

```diff
@@ -56,9 +56,17 @@ TraceCollector::~TraceCollector()
   */
 void TraceCollector::stop()
 {
-    WriteBufferFromFileDescriptor out(TraceSender::pipe.fds_rw[1]);
-    writeChar(true, out);
-    out.next();
+    try
+    {
+        WriteBufferFromFileDescriptor out(TraceSender::pipe.fds_rw[1]);
+        writeChar(true, out);
+        out.next();
+    }
+    catch (...)
```

Comment:
`TraceCollector::stop() ` is called only from `~TraceCollector`. I think we can combine them together, to have a single try/catch.

Also probably move `TraceSender::pipe.close();` and `thread.join();` into `SCOPE_EXIT_SAFE`